clippy_lints/src/casts/cast_possible_truncation.rs

```diff
@@ -159,5 +161,18 @@ pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>, cast_expr: &Expr<'_>,
         _ => return,
     };
 
-    span_lint(cx, CAST_POSSIBLE_TRUNCATION, expr.span, &msg);
+    let mut applicability = Applicability::MachineApplicable;
+
+    let snippet = snippet_with_applicability(cx, expr.span, "x", &mut applicability);
+    let name_of_cast_from = snippet.split(" as").next().unwrap_or("x");
+
+    span_lint_and_sugg(
+        cx,
+        CAST_POSSIBLE_TRUNCATION,
+        expr.span,
+        &msg,
+        "avoid silent truncation by using",
+        format!("{cast_to}::try_from({name_of_cast_from}).unwrap()"),
```

Comment:
I think using using `expect` here and hinting developers to explain why it shouldn't truncate would be even better
```suggestion
        format!("{cast_to}::try_from({name_of_cast_from}).expect("truncation should be impossible because ...)"),
```

Reply:
I've made this change and changed MachineApplicable to HasPlaceholders, is this the best way to classify this?

I tried to come up with some sort of `u32::try_from({} & {cast_to}::MAX as {cast_from}).unwrap()` machine applicable suggestion, but they all bury the intention of the lint.

Reply:
`HasPlaceholders` is fine here 👍 Another option now that I think about it would be to make the whole `expect` reason a placeholder so that the user wouldn't be able to copy-paste the offered suggestion without having to think about the issue before getting a compilable version of their program. What do you think @xFrednet? 
```rust
format!("{cast_to}::try_from({name_of_cast_from}).expect(..)
```

Reply:
Thank you for the review and feedback @kraktus, it's highly appreciated!

Using `try_from()` changes the semantics of the code. In most cases, users would most likely prefer the value to be truncated over the code panicking in production. This also misses cases, where this behavior might be intentional. Though, looking at the lint description, it's not ideal to emit a lint without any suggestion.

The idea of using `try_from` is good and should be included in the lint message. The applicability of `HasPlaceholders` seamed a bit high to me at first, but good at a second thought.

That being said, I would suggest a different lint message. Something like this:

```rs
error: casting `f64` to `isize` may truncate the value
  --> $DIR/cast.rs:29:5
   |
LL |     1f64 as isize;
   |
help: if this is intensional allow the lint with `#[allow(clippy::cast_precision_loss)]` ...
help: ... or use `try_from` and handle the error accordingly
   |
LL | isize::try_from(1f64)
   |
```

Note that this changes the semantics even more, as we don't instruct the user to use `.expect("..")` or `.unwrap()`. For this, I would make the applicability unspecified.

What are your thoughts about this? :upside_down_face: 

---

You can create more complex messages with: `span_lint_and_then`.

<details>
<summary>Here is a mockup how the code could look:</summary>
```rs
span_lint_and_then(
    cx,
    <LINT>,
    <span>,
    <msg>,
    |diag| {
        diag.help(
            "if this is intensional allow the lint with `#[allow(clippy::cast_precision_loss)]` ...",
        );
        diag.span_suggestion_with_style(
            <span>,
            "... or use `try_from` and handle the error accordingly",
            suggestion,
            Applicability::Unspecified,
            // always show the suggestion in a separate line
            SuggestionStyle::ShowAlways,
        );
    },
);
``` 
</details>

Reply:
This is correct, i mixed up the formatting. You could try add the #[allow] attribute as another code suggestion, but I think this is better. Only thing is the note in the middle which we sadly can influence directly. 

There is sadly no detailed documentation for the formatting.

Could you also update the lint description to mention these two solutions?

Reply:
Yes, I put 'why is this bad' back to how it was, and appended a bit to the 'what it does' to reflect both the ignore and try_from suggestions.